XenAdmin/SettingsPanels/NRPEEditPage.cs

```diff
@@ -0,0 +1,394 @@
+            }
+
+            AllowHostsTextBox.Text = NRPEHostConfiguration.ALLOW_HOSTS_PLACE_HOLDER;
+            AllowHostsTextBox.ForeColor = Color.FromKnownColor(KnownColor.ControlDark);
+            AllowHostsTextBox.GotFocus += AllowHostsTextBox_GotFocus;
+            AllowHostsTextBox.LostFocus += AllowHostsTextBox_LostFocus;
+
+            _invalidParamToolTip = new ToolTip
+            {
+                IsBalloon = true,
+                ToolTipIcon = ToolTipIcon.Warning,
+                ToolTipTitle = Messages.INVALID_PARAMETER,
+                Tag = AllowHostsTextBox
+            };
+        }
+
+        public string SubText => Messages.NRPE_EDIT_PAGE_TEXT;
+
+        public bool ValidToSave
+        {
+            get
+            {
+                _invalidParamToolTipText = "";
+                _invalidParamToolTip.ToolTipTitle = "";
+
+                if (!EnableNRPECheckBox.Checked)
+                {
+                    return true;
+                }
+
+                bool valid = IsAllowHostsValid();
+
+                DataGridViewTextBoxCell focusCellWhenErrorOccurs = null;
+                foreach (CheckGroup checkGroup in _checkGroupList)
+                {
+                    if (!checkGroup.CheckValue())
+                    {
+                        valid = false;
+                        if (focusCellWhenErrorOccurs == null)
+                        {
+                            focusCellWhenErrorOccurs = checkGroup.NameCell;
+                        }
+                    }
+                }
+                if (focusCellWhenErrorOccurs != null)
+                {
+                    CheckDataGridView.CurrentCell = CheckDataGridView.Rows[0].Cells[0];
+                    CheckDataGridView.CurrentCell = focusCellWhenErrorOccurs;
+                }
+                return valid;
+            }
+        }
+
+        public bool HasChanged
+        {
+            get
+            {
+                UpdateCurrentNRPEConfiguration();
+                return !_nrpeCurrentConfig.Equals(_nrpeOriginalConfig);
+            }
+        }
+
+        public void Cleanup()
+        {
+        }
+
+        public void ShowLocalValidationMessages()
+        {
+            if (_invalidParamToolTip.Tag is Control ctrl)
+            {
+                _invalidParamToolTip.Hide(ctrl);
+                if (!_invalidParamToolTipText.Equals(""))
+                {
+                    HelpersGUI.ShowBalloonMessage(ctrl, _invalidParamToolTip, _invalidParamToolTipText);
+                }
+            }
+        }
+
+        public void HideLocalValidationMessages()
+        {
+            if (_invalidParamToolTip.Tag is Control ctrl)
+            {
+                _invalidParamToolTip.Hide(ctrl);
+            }
+        }
+
+        public AsyncAction SaveSettings()
+        {
+            return new NRPEUpdateAction(_clone, _nrpeCurrentConfig, _nrpeOriginalConfig, true);

```

Comment:
I didn't spot this earlier: why are we suppressing the history? This means that there is not record of the action on the events TabPage. Same for line 195 and NRPERetrieveAction.

Reply:
Hmm, actually we do suppress this for more properties that we save. It probably needs revisiting at a higher level, so it is probably a no-action for this PR.